addons/better-terrain/editor/Dock.gd

```diff
@@ -314,57 +333,74 @@ func _on_remove_terrain_pressed() -> void:
 		undo_manager.commit_action()
 
 
-func perform_add_terrain(name: String, color: Color, type: int, categories: Array) -> void:
-	if BetterTerrain.add_terrain(tileset, name, color, type, categories):
-		var new_terrain = terrain_tree.create_item(terrain_tree.get_root())
-		new_terrain.set_text(0, name)
-		new_terrain.set_icon(0, terrain_icons[type])
-		new_terrain.set_icon_modulate(0, color)
+func add_terrain_entry(terrain:Dictionary, index:int = -1):
+	if index < 0:
+		index = terrain_list.get_child_count()
+	
+	var entry = TERRAIN_ENTRY_SCENE.instantiate()
+	entry.tileset = tileset
+	entry.terrain = terrain
+	entry.select.connect(_on_entry_select)
+	
+	terrain_list.add_child(entry)
+
+
+func remove_terrain_entry(index:int):
+	terrain_list.get_child(index).free()
+
+
+func perform_add_terrain(name: String, color: Color, type: int, categories: Array, icon:Dictionary = {}) -> void:
+	if BetterTerrain.add_terrain(tileset, name, color, type, categories, icon):
+		var index = BetterTerrain.terrain_count(tileset) - 1
+		var terrain = BetterTerrain.get_terrain(tileset, index)
+		add_terrain_entry(terrain, index)
 
 
 func perform_remove_terrain(index: int) -> void:
-	var root = terrain_tree.get_root()
-	if index >= root.get_child_count():
+	if index >= terrain_list.get_child_count():
 		return
-	var item = root.get_child(index)
+	var item = terrain_list.get_child(index)
 	if BetterTerrain.remove_terrain(tileset, index):
-		item.free()
+		remove_terrain_entry(index)
 		update_tile_view_paint()
 
 
 func perform_swap_terrain(index1: int, index2: int) -> void:
 	var lower := min(index1, index2)
 	var higher := max(index1, index2)
-	var root = terrain_tree.get_root()
-	if lower >= root.get_child_count() or higher >= root.get_child_count():
+	if lower >= terrain_list.get_child_count() or higher >= terrain_list.get_child_count():
 		return
-	var item1 = root.get_child(lower)
-	var item2 = root.get_child(higher)
+	var item1 = terrain_list.get_child(lower)
+	var item2 = terrain_list.get_child(higher)
 	if BetterTerrain.swap_terrains(tileset, lower, higher):
-		item2.move_before(item1)
-		item1.move_after(root.get_child(higher))
+		terrain_list.move_child(item1, higher)
+		selected_entry = index2
+		terrain_list.get_child(index2).set_selected(true)
 		update_tile_view_paint()
 
 
-func perform_edit_terrain(index: int, name: String, color: Color, type: int, categories: Array) -> void:
-	var root = terrain_tree.get_root()
-	if index >= root.get_child_count():
+func perform_edit_terrain(index: int, name: String, color: Color, type: int, categories: Array, icon: Dictionary = {}) -> void:
+	if index >= terrain_list.get_child_count():
 		return
-	var item = root.get_child(index)
-	if BetterTerrain.set_terrain(tileset, index, name, color, type, categories):
-		item.set_text(0, name)
-		item.set_icon(0, terrain_icons[type])
-		item.set_icon_modulate(0, color)
+	var entry = terrain_list.get_child(index)
+	# don't overwrite empty icon
+	var _icon = icon
```

Comment:
Godot uses initial underscore to mark unused vars. I suggest changing `_icon` to `valid_icon` or something like that.

Reply:
Oh, it also does it for private vars doesn't it? Boy, I'm not thinking clearly today :rofl: Up to you if you want to give it a nice name

Reply:
Yeah I tend to use this convention for local/private vars. I've renamed it either way :)